# CityBike Wien Datensetanalye für die Jahre 2016 und 2017

== Inhalt und Struktur dieses Notebooks 



## Datenset importieren bereinigen und transformieren

Die Daten werden zuerst eingelesen und nötige Transformationen werden vorgenommen. 

In erste Linie werden alle Fahrten die unter zwei Minuten dauern und an der gleichen Station enden aus dem Datenset herausgenommen da man davon ausgehen kann, dass der Nutzer das System nicht benutzt hat. 

In [2]:
%matplotlib inline
import pandas as pd
import datetime
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import math
#Warnungen seitens Jupyter sollen ignoriert werden
warnings.filterwarnings('ignore')

df2016 = pd.read_csv('CityBikeWien/2016trips.csv', error_bad_lines=False, parse_dates=['Startdatum', 'Enddatum'], delimiter=';')
df2017 = pd.read_csv('CityBikeWien/2017trips.csv', error_bad_lines=False, parse_dates=['Startdatum', 'Enddatum'], delimiter=';')

Die beiden DataFrames für die Jahre 2016 und 2017 werden zusammengeführt

In [3]:
df_total = df2016.append(df2017)

Bei dem neuen DataFrame wird der Index neu gesetzt

In [4]:
df_total = df_total.reset_index(drop=True)

Eine neue Spalte für die Fahrtdauer wird hinzugefügt (Enddatum-Startdatum)

In [5]:
df_total['Fahrtdauer'] =  df_total['Enddatum'] - df_total['Startdatum']

Uns interessieren nur Fahrten die länger als zwei Minuten dauern und gleichzeitig nicht an der gleichen Station beendet wurden. Die Bereinigung wird für alle dreit Datasets vorgenommen.

In [19]:
df_total_cleansed = df_total[ (df_total['Fahrtdauer'] > pd.Timedelta('120 sec')) 
                            & (df_total['Anfangsstation_ID'] != df_total['Endstation_ID'])]

df2016            = df2016[ (df2016['Fahrtdauer'] > pd.Timedelta('120 sec'))
                            & (df2016['Anfangsstation_ID'] != df2016['Endstation_ID'])]

df2017            = df2017[ (df2017['Fahrtdauer'] > pd.Timedelta('120 sec'))
                            & (df2017['Anfangsstation_ID'] != df2017['Endstation_ID'])]

Das bereinigte DataFrame enthält insgesamt 1.897.764 Einträge für die Jahre 2016 und 2017 mit folgenden Datentypen

In [20]:
df_total_cleansed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1897764 entries, 1 to 2080071
Data columns (total 7 columns):
Startdatum           datetime64[ns]
Enddatum             datetime64[ns]
Anfangsstation_ID    int64
Endstation_ID        int64
Fahrrad_ID           int64
Subscriber_ID        object
Fahrtdauer           timedelta64[ns]
dtypes: datetime64[ns](2), int64(3), object(1), timedelta64[ns](1)
memory usage: 115.8+ MB


Erste Zeile des DataFrames als Referenz

In [21]:
df_total_cleansed.head(1)

Startdatum            Enddatum  Anfangsstation_ID  Endstation_ID  \
1 2016-01-01 00:13:04 2016-01-01 00:38:05               1026           1100   

   Fahrrad_ID                       Subscriber_ID Fahrtdauer  
1        2539  0x04084286D0CC7FA3973C16F8306802A1   00:25:01

## Finde die ersten 10 Benutzer nach Benutzung für den gesamten Zeitraum (2016+2017)


In der folgenden Zelle wird das DataFrame anhand der 'Subscriber_ID' gruppiert. 

Darauffolgend wird eine Aggregatfunktion (np.size) auf das gruppierte Objekt angewendet um die Anzahl der Fahrten je Subscriber zu erhalten. Somit erhält man ein Series-Objekt welches sortiert werden muss. 

Um die Handhabung zu erleichtern wird das Series-Objekt zu einem DataFrame mit einer einzigen Spalte transformiert.

In [22]:
grouped_by_subscriber_groups              = df_total.groupby('Subscriber_ID')
aggregated_by_subscriber_series           = grouped_by_subscriber_groups['Fahrrad_ID'].agg(np.size)
aggregated_sorted_by_subscriber_series    = aggregated_by_subscriber_series.sort_values(ascending=False)
aggregated_by_subscriber_dataframe        = aggregated_sorted_by_subscriber_series.to_frame('ANZAHL_FAHRTEN')

Die ersten zehn Subscriber werden extrahiert und zudem wird eine Spalte mit dem Rang hinzugefügt. Danach wird das DataFrame ausgegeben. 

In [23]:
aggregated_by_subscriber_dataframe_top_10 = aggregated_by_subscriber_dataframe[:10]
aggregated_by_subscriber_dataframe_top_10['RANG'] = np.arange(1,11) 
aggregated_by_subscriber_dataframe_top_10

ANZAHL_FAHRTEN  RANG
Subscriber_ID                                           
0x30EF7C4D0C1211B47ADBA3295EF47C61            1790     1
0xB9AF912E74C4F6602CF849721E28B621            1400     2
0xAC44D643FD2F8AE0CCD64B35A269DCC1            1297     3
0x30F9A85D6ADB8CFA380C4EA7D77C574E            1281     4
0xCA316F5E99321FD79D33C3CAB1000DA5            1193     5
0xE88FAEE35C9817B6BD91961D8DD186E0            1028     6
0x62B4268B5D5AB0FD8B0FDC93D79498DD            1025     7
0x24384492DBD8ACA63F4C29FC8BC3B36A            1011     8
0xD372854C5186AB811AEE55325547CC85            1009     9
0x0156CCFCAAF6097587BCB2B1C711F1FE             980    10

## Was hat der Benutzer mit den meisten Fahrten für die Nutzung des Systems bezahlt?

Der 'Citybike Wien'-Nutzer welcher das System innerhalb der beiden Jahren am meisten genutzt hat, fuhr insgesamt 1790 Fahrten. Es wäre interessant zu sehen was ihn die Nutzung des Systems über die Zeitdauer der beiden Jahre gekostet hat. 

Hierfür ist es nötig den Kostenplan von Citybike Wien zu kennen. Dieser ist progressiv eingestellt, das heißt, dass die erste Stunde kostenlos ist. Die zweite Stunde kostet einen Euro, die dritte das doppelte und die vierte Stunde dann vier Euro. Jede weitere Stunde kostet jeweils vier Euro. 

Wir suchen alle Fahrten für den Power-User heraus

In [24]:
df_user_1 = df_total[df_total['Subscriber_ID'] == '0x30EF7C4D0C1211B47ADBA3295EF47C61']

Wie oft hat der Nutzer das Limit der einen Stunde überschritten? Insgesamt 98 mal.

In [25]:
df_user_1[df_user_1['Fahrtdauer'] > pd.Timedelta('60 min')].size

98

Gab es Fälle bei denen die Fahrt länger als zwei Stunden dauerte? Nein

In [26]:
df_user_1[df_user_1['Fahrtdauer'] > pd.Timedelta('120 min')].size

0

Insgesamt fielen 98 kostenpflichtige Fahrten an welche weniger als zwei Stunden gedauert haben. Somit hat der Anwender insgesamt 98 Euro für die Nutzung des Systems für die Zeitdauer der beiden Jahren bezahlt da jede Fahrt genau einen Euro gekostet hat.

Im Schnitt hat der oder die NutzerIn ~ 5,5 Cent für jede Fahrt bezahl.


## Welchen Umsatz hat das System für das Jahr 2016 bzw. 2017 generiert?

Es wäre interessant zu sehen wie viel Umsatz alleine durch die Nutzung des Systems entstanden ist. 

Der Betreiber von CityBike Wien, die Gewista GmbH wählt ein hybrides Finanzierungsmodell. Das heißt, dass sich das System nicht nur aus den Beiträgen der einzelnen Nutzer finanziert, sondern auch aus Werbungen welche auf den Fahrrädern angezeigt werden. Inwieweit oder ob das System durch die Stadt Wien subventioniert wird ist nicht bekannt. 

Wir untersuchen nun die Einnahmen welche durch die Nutzergemeinde des Systems entsteht. Dies ist durch die Kenntniss des Kostenmodells und der Länge der einzelnen Fahrten möglich. Die Anmeldegebühr von einem Euro wird nicht mit in die Berechnung einbezogen da aus dem Datenset nicht bekannt ist wann sich der Nutzer das erste Mal registriert hat. Somit könnte es sein, dass die Anmeldegebühr nicht in den Jahren 2016 oder 2017 verrechnet wurde. 

In [27]:
df2016['Fahrtdauer'] = df2016['Enddatum'] - df2016['Startdatum']
df2017['Fahrtdauer'] = df2017['Enddatum'] - df2017['Startdatum']

In [79]:
def calculate_fare(fahrtdauer):
    ssddsg=  math.ceil(fahrtdauer.total_seconds() // 3600)
    if(fahrtdauer < pd.Timedelta('60 min')):        return int(0)
    elif( (fahrtdauer > pd.Timedelta('60 min')) & (fahrtdauer < pd.Timedelta('120 min'))):
        return int(1)
    elif( (fahrtdauer > pd.Timedelta('120 min')) & (fahrtdauer < pd.Timedelta('180 min'))):
        return int(3)
    elif( (fahrtdauer > pd.Timedelta('180 min')) & (fahrtdauer < pd.Timedelta('240 min'))):
        return int(7)
    elif((fahrtdauer > pd.Timedelta('240 min'))):
        return (int(7) +  ((math.ceil(fahrtdauer.total_seconds() // 3600))-3)*4)
    #a case where the bike is not returned within six days 

In [80]:
df2016_test = df2016

In [81]:
df2016_test['Fahrtpreis'] = df2016_test['Fahrtdauer'].apply(calculate_fare)

In [82]:
df2016_test[ df2016_test['Fahrtdauer'] > pd.Timedelta('21000 sec') ]

Startdatum                Enddatum  Anfangsstation_ID  \
71      2016-01-01 05:03:33 2016-01-01 12:05:51.000               1046   
77      2016-01-01 05:58:09 2016-01-01 13:50:14.000               1024   
5621    2016-01-11 22:08:35 2016-01-12 04:14:24.000               1063   
10794   2016-01-15 22:12:55 2016-01-16 10:58:06.000               1076   
11548   2016-01-16 18:59:58 2016-01-17 00:59:53.000               1086   
11798   2016-01-17 02:07:38 2016-01-17 12:59:27.000               1116   
11919   2016-01-17 10:12:40 2016-01-17 16:48:10.000               1125   
13618   2016-01-19 09:46:34 2016-01-19 23:47:32.000               1044   
19428   2016-01-26 15:04:02 2016-01-27 12:39:51.000               1146   
22488   2016-01-28 12:19:22 2016-01-28 18:39:49.000               1134   
22498   2016-01-28 12:24:09 2016-01-28 18:39:40.000               1134   
22502   2016-01-28 12:29:07 2016-01-28 18:39:49.000               1134   
29061   2016-02-02 14:16:36 2016-02-03 08:39:29.000               1128   
30373   2016-02-03 08:54:26 2016-02-03 21:27:36.000               1105   
35199   2016-02-07 10:35:08 2016-02-07 20:39:01.000               1068   
35682   2016-02-07 14:12:54 2016-02-08 07:11:20.000               1106   
38263   2016-02-08 21:03:05 2016-02-09 06:56:03.000               1068   
41173   2016-02-11 09:03:14 2016-02-11 18:15:46.000               1046   
42212   2016-02-12 00:19:42 2016-02-12 06:58:50.000               1024   
42454   2016-02-12 09:37:21 2016-02-12 16:41:44.000               1045   
44035   2016-02-13 13:49:02 2016-02-14 13:08:28.000               1024   
44060   2016-02-13 13:55:06 2016-02-15 09:28:22.000               1049   
44207   2016-02-13 14:42:17 2016-02-14 14:09:27.000               1129   
44657   2016-02-13 17:46:45 2016-02-14 11:20:32.000               1031   
44660   2016-02-13 17:48:18 2016-02-14 11:20:36.000               1031   
45215   2016-02-14 10:02:23 2016-02-14 16:48:31.000               1078   
45423   2016-02-14 12:55:26 2016-02-15 07:21:34.000               1056   
46229   2016-02-15 10:50:03 2016-02-16 12:50:41.000               1029   
51460   2016-02-20 01:29:18 2016-02-20 12:14:26.000               1116   
51897   2016-02-20 13:07:49 2016-02-20 20:41:33.000               1069   
...                     ...                     ...                ...   
1027271 2016-11-23 23:03:00 2016-11-24 07:24:16.000               1129   
1028884 2016-11-25 01:13:59 2016-11-25 12:08:36.000               1032   
1030222 2016-11-26 01:24:26 2016-11-27 14:14:14.000               1092   
1030602 2016-11-26 10:50:09 2016-11-26 17:04:18.000               1033   
1040525 2016-12-03 21:16:30 2016-12-05 10:08:17.000               1021   
1041417 2016-12-04 19:41:34 2016-12-05 07:27:29.000               1076   
1042800 2016-12-06 08:18:22 2016-12-06 15:05:20.000               1114   
1046366 2016-12-09 10:46:15 2016-12-09 17:11:10.000               1029   
1047451 2016-12-10 00:50:07 2016-12-10 09:17:51.000               1095   
1047529 2016-12-10 01:44:35 2016-12-10 09:57:28.000               1136   
1047537 2016-12-10 01:56:30 2016-12-12 16:58:52.250               1054   
1047809 2016-12-10 10:46:11 2016-12-10 17:46:57.000               1035   
1047821 2016-12-10 10:51:12 2016-12-10 17:47:41.000               1035   
1047844 2016-12-10 10:58:46 2016-12-10 17:47:14.000               1035   
1049564 2016-12-11 12:57:53 2016-12-12 11:23:46.000               1075   
1049597 2016-12-11 13:27:22 2016-12-11 20:25:14.000               1045   
1049621 2016-12-11 13:51:16 2016-12-12 07:28:09.000               1080   
1049761 2016-12-11 15:54:40 2016-12-12 01:42:13.000               1071   
1054861 2016-12-15 23:15:11 2016-12-16 06:50:08.000               1071   
1056276 2016-12-16 23:58:18 2016-12-17 16:46:23.000               1138   
1056652 2016-12-17 12:33:11 2016-12-17 22:31:45.000               1106   
1058842 2016-12-19 23:50:11 2016-12-20 07:06:47.000               1061   
1060197 2016-12-21 22